drivers/sensor/ti/ina7xx/ina7xx.c

```diff
@@ -0,0 +1,394 @@
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_POWER_BIT, true);
+	}
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_CURRENT) {
+
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_CURRENT_BIT, false);
+		rc = ina7xx_read16(dev, INA7xx_REG_CURRENT, &data->current);
+		if (rc < 0) {
+			LOG_ERR("Error reading current register");
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_CURRENT_BIT, true);
+	}
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_DIE_TEMP) {
+
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_DIE_TEMP_BIT, false);
+		rc = ina7xx_read16(dev, INA7xx_REG_DIE_TEMP, &data->die_temp);
+		if (rc < 0) {
+			LOG_ERR("Error reading temperature register");
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_DIE_TEMP_BIT, true);
+	}
+
+	return rc;
+}
+
+static int ina7xx_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct ina7xx_data *data = dev->data;
+	int16_t signed_current;
+	int16_t signed_die_temp;
+
+	switch (chan) {
+	case SENSOR_CHAN_VOLTAGE:
+		if (IS_BIT_SET(data->valid_readings_mask, INA7xx_READ_VOLTAGE_BIT)) {
+			sensor_value_from_micro(val,
+						(int64_t)data->v_bus * INA7xx_BUS_VOLTAGE_MUL_UV);
+		} else {
+			LOG_WRN("Last fetch did not include a voltage reading");
+			return -ENODATA;
+		}
+		break;
+	case SENSOR_CHAN_DIE_TEMP:
+		if (IS_BIT_SET(data->valid_readings_mask, INA7xx_READ_DIE_TEMP_BIT)) {
+			signed_die_temp = (int16_t)data->die_temp;
+			signed_die_temp /= 16; /* Remove RESERVED bits 3-0 */
+			sensor_value_from_micro(val, signed_die_temp * INA7xx_TEMP_SCALE_MG);
+		} else {
+			LOG_WRN("Last fetch did not include a die temp reading");
+			return -ENODATA;
+		}
+		break;
+	case SENSOR_CHAN_POWER:
+		if (IS_BIT_SET(data->valid_readings_mask, INA7xx_READ_POWER_BIT)) {
+			sensor_value_from_micro(val, (int64_t)data->power * data->power_mul_uw);
+		} else {
+			LOG_WRN("Last fetch did not include a power reading");
+			return -ENODATA;
+		}
+		break;
+	case SENSOR_CHAN_CURRENT:
+		if (IS_BIT_SET(data->valid_readings_mask, INA7xx_READ_CURRENT_BIT)) {
+			signed_current = (int16_t)data->current;
+			sensor_value_from_micro(val,
+						signed_current * (int64_t)data->current_mul_ua);
+		} else {
+			LOG_WRN("Last fetch did not include a current reading");
+			return -ENODATA;
+		}
+		break;
+	default:
+		LOG_DBG("Channel not supported by device");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ina7xx_init(const struct device *dev)
+{
+	const struct ina7xx_config *cfg = dev->config;
+	struct ina7xx_data *data = dev->data;
+	int rc;
+	uint16_t tmp;
+
+	if (!i2c_is_ready_dt(&cfg->bus)) {
+		LOG_ERR("Device not ready");
+		return -ENODEV;
+	}
+
+	rc = ina7xx_write(dev, INA7xx_REG_CONFIG, INA7xx_RST_MASK << INA7xx_RST_SHIFT);
+	if (rc < 0) {
+		LOG_ERR("Could not reset device");
+		return rc;
+	}
+
+	/* 300us reset time */
+	k_sleep(K_USEC(INA7xx_POWERUP_USEC));
```

Comment:
Note if we add the conditional delay INA7xx_POWERUP_USEC it needs to be before I2C access:

```suggestion
	k_sleep(K_TIMEOUT_ABS_US(INA7xx_POWERUP_USEC);

	rc = ina7xx_write(dev, INA7xx_REG_CONFIG, INA7xx_RST_MASK << INA7xx_RST_SHIFT);
	if (rc < 0) {
		LOG_ERR("Could not reset device");
		return rc;
	}
```

HI @Schulle73 if I understand correctly because the init function is currently also the resume from sleep there are two different power up / resume delay requirements. A possible solution would be to do the following:

Rename the ina7xx_init() to ina7xx_config() or similar and then create two new functions as so:

```
static int ina7xx_init(const struct device *dev)
{
	k_sleep(K_TIMEOUT_ABS_US(INA7xx_POWERUP_USEC);

	ina7xx_config(dev);
}
```

and 

```
static int ina7xx_pm_resume(const struct device *dev)
{
	k_sleep(K_USEC(INA7xx_WAIT_STARTUP_USEC);

	ina7xx_config(dev);
}
```
Other than this suggestion if others are happy I'm ready to approve.

Reply:
Hi @nixward, i'm in discussion with TI about that. I hope the I2C is't blocked for INA7xx_POWERUP_USEC. And and I want to wait for the opinion of @yperess. He demanded changes that he hadn't yet reviewed. And that's blocking PR. And he wasn't a friend of delays in the init phase.

Reply:
you don't need the ina7xx_pm_resume function. the device goes to the shutdown mode if you triggered a conversion. you have to wait INA7xx_WAIT_STARTUP_USEC every fetch if you use the triggered conversion modes. that's why im add this to the conversion time.
The pm_action function only make sense it the device runs in continuous mode. I have to look at this again.